### PR TITLE
feat(component): add SideNavItems.tsx component

### LINKED ISSUE #13603: Add TypeScript types to SideNavItems

### PR DESCRIPTION
**Pull request by Nirajsah**
Closes #13603

Detailed Description

- In this pull request, I have introduced a new component file, SideNavItems.tsx, along with corresponding TypeScript types to enhance type safety. During the development of this component, I encountered two issues. first issue was related to propTypes, which I removed in earlier commits i found that it is required. Additionally, there was an issue related to TypeScript's ability to correctly infer types, particularly with regards to (child.type?.displayName).

To resolve the issue with TypeScript's type inference, I implemented a solution by explicitly defining the child's type as React.ComponentType<any>. I then checked whether the displayName property existed within childType. This approach ensured that TypeScript correctly recognized and inferred the types, effectively resolving the issue.

In previous PRs, I unintentionally omitted running the tests before making the commits. I acknowledge and apologise for this oversight.

New

- Added a new file SideNavItems.tsx along with its corresponding TypeScript types. Issue Resolution

Changed
- None

Removed
- None

It's important to note that all other aspects of the SideNavItems.tsx file are functioning as expected, and there are no other issues in the codebase.

Testing / Review
Reviewers can verify the changes in this PR by:

- Examining the new SideNavItems.tsx file and its associated TypeScript types for correctness and completeness.
- Confirming that the issue related to the isSideNavExpanded prop has been successfully resolved.
- Ensuring that the removal of the SideNavItems.propTypes block does not introduce any new issues, as TypeScript now handles type checking.

### CODE REVIEW

**Comment by Nirajsah** [comment]
ci / test:e2e failed running "yarn install --immutable --immutable-cache"

**Comment by Nirajsah** [comment]
I closed and then reopened the PR, and this time, all tests passed successfully. It's possible that something may have gone wrong during the initial PR's test run, but the subsequent attempt resulted in the expected test outcomes.

**Review by guidari — APPROVED** [review]
LGTM! 🚀 Thanks for the contribution!

File: packages/react/src/components/UIShell/SideNavItems.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  //    * Provide an optional class to be applied to the containing node
+  //    */
+  className: PropTypes.string,
+
+  //   /**
+  //    * Property to indicate if the side nav container is open (or not). Use to
+  //    * keep local state and styling in step with the SideNav expansion state.
+  //    */
+  isSideNavExpanded: PropTypes.bool,
+};
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/SideNavItems.tsx]: Just a minor formatting nit here

```suggestion
SideNavItems.displayName = 'SideNavItems';
SideNavItems.propTypes = {
  /**
   * Provide a single icon as the child to `SideNavIcon` to render in the
   * container
   */
  children: PropTypes.node.isRequired,

  /**
   * Provide an optional class to be applied to the containing node
   */
  className: PropTypes.string,

  /**
   * Property to indicate if the side nav container is open (or not). Use to
   * keep local state and styling in step with the SideNav expansion state.
   */
  isSideNavExpanded: PropTypes.bool,
};
```

File: packages/react/src/components/UIShell/SideNavItems.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * Provide a single icon as the child to `SideNavIcon` to render in the
+   * container, and it is required
+   */
+  children: React.ReactNode;
+  /**
+   * Property to indicate if the side nav container is open (or not). Use to
+   * keep local state and styling in step with the SideNav expansion state.
+   */
+  isSideNavExpanded?: boolean;
```
Comment (tay1orjones) [inline packages/react/src/components/UIShell/SideNavItems.tsx]: Just some formatting pieces here

```suggestion
  /**
   * Provide a single icon as the child to `SideNavIcon` to render in the
   * container
   */
  children: React.ReactNode;

  /**
   * Provide an optional class to be applied to the containing node
   */
  className?: string;

  /**
   * Property to indicate if the side nav container is open (or not). Use to
   * keep local state and styling in step with the SideNav expansion state.
   */
  isSideNavExpanded?: boolean;
```

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Thanks for submitting this PR! The `.tsx` file should be replacing the `.js` one though. There's some [docs here](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/adding-component-types.md#steps-to-provide-baseline-type-definitions-for-components) on the steps we're taking for each of these components if that's helpful.

File: packages/react/src/components/UIShell/SideNavItems.tsx (hunk 1 above)
Comment (Nirajsah) [reply packages/react/src/components/UIShell/SideNavItems.tsx]: ohh yes, i'll fix it right away

Comment (Nirajsah) [reply packages/react/src/components/UIShell/SideNavItems.tsx]: sorry about that🥲

**Comment by Nirajsah** [comment]
> Thanks for submitting this PR! The `.tsx` file should be replacing the `.js` one though. There's some [docs here](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/adding-component-types.md#steps-to-provide-baseline-type-definitions-for-components) on the steps we're taking for each of these components if that's helpful.

I've reviewed the documentation and will bear it in mind.

File: packages/react/src/components/UIShell/SideNavItems.tsx (hunk 1 above)
Comment (Nirajsah) [reply packages/react/src/components/UIShell/SideNavItems.tsx]: done.🚀

File: packages/react/src/components/UIShell/SideNavItems.tsx (hunk 2 above)
Comment (Nirajsah) [reply packages/react/src/components/UIShell/SideNavItems.tsx]: i'll keep formatting in mind before making a PR.

**Comment by Nirajsah** [comment]
some checks failed i don't know why, closing the PR. I'll try to reopen the PR for checks

**Comment by Nirajsah** [comment]
closing this PR. I'll Make a new PR after everything thing is fixed

**Pull request closed without merging**